### PR TITLE
[Draft] Integrate ImpactIA methodology

### PR DESCRIPTION
**Pull request by benoit-cty**
<h2 aria-level="5" style="unicode-bidi: plaintext; color: rgb(204, 204, 204); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Roboto, sans-serif; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(37, 37, 38); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">What Claiude Fable 5.1 did</h2><p style="white-space: pre-wrap; margin-top: 0.1em; margin-bottom: 0.2em; unicode-bidi: plaintext; color: rgb(204, 204, 204); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Roboto, sans-serif; font-size: 13px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(37, 37, 38); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">The ImpactIA work is now integrated on the <code style="font-family: monospace; color: rgb(215, 186, 125); background: none 0% 0% / auto repeat scroll padding-box border-box rgba(255, 255, 255, 0.1); padding: 2px 4px; border-radius: 3px; word-break: break-word; font-size: 0.9em;">feat/impactia</code> branch (uncommitted, 27 files). Lint passes, the docs build, and the suite is at 211 passed. The 4 failures in <code style="font-family: monospace; color: rgb(215, 186, 125); background: none 0% 0% / auto repeat scroll padding-box border-box rgba(255, 255, 255, 0.1); padding: 2px 4px; border-radius: 3px; word-break: break-word; font-size: 0.9em;">tests/test_openai.py</code> (Responses API) fail identically on the untouched branch: the recorded cassette lacks <code style="font-family: monospace; color: rgb(215, 186, 125); background: none 0% 0% / auto repeat scroll padding-box border-box rgba(255, 255, 255, 0.1); padding: 2px 4px; border-radius: 3px; word-break: break-word; font-size: 0.9em;">cache_write_tokens</code>, which the current OpenAI SDK requires, so they are unrelated.</p><p style="white-space: pre-wrap; margin-top: 0.1em; margin-bottom: 0.2em; unicode-bidi: plaintext; color: rgb(204, 204, 204); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Roboto, sans-serif; font-size: 13px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(37, 37, 38); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><strong>Core inference model, on by default</strong> (<a href="vscode-webview://13qe2srtq7rfa1e0j47ttranh8f06vadkqqo1rgicmqsrna8gimk/ecologits/impacts/llm.py" target="_blank" rel="noopener noreferrer" style="color: rgb(55, 148, 255); text-decoration: rgb(55, 148, 255);">llm.py</a>):</p><ul style="padding-inline-start: 2em; color: rgb(204, 204, 204); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Roboto, sans-serif; font-size: 13px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(37, 37, 38); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><li style="unicode-bidi: plaintext;"><strong>Input tokens now matter.</strong><span> </span>Pre-fill latency is modeled from the input token count (Agrawal et al. coefficient, scaled by each model's OpenRouter TTFT, as ImpactIA proposes). All seven tracers pass the provider-reported input tokens.</li><li style="unicode-bidi: plaintext;"><strong>Datacenter network equipment</strong><span> </span>(firewall, router, switch): energy plus embodied GWP, allocated like the server.</li><li style="unicode-bidi: plaintext;"><strong>Datacenter building</strong><span> </span>embodied GWP factor per kWh of IT energy.</li><li style="unicode-bidi: plaintext;">ADPe and PE are not extended for these two items because ImpactIA only provides GWP data; the docs say so.</li></ul><p style="white-space: pre-wrap; margin-top: 0.1em; margin-bottom: 0.2em; unicode-bidi: plaintext; color: rgb(204, 204, 204); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Roboto, sans-serif; font-size: 13px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(37, 37, 38); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><strong>New experimental <code style="font-family: monospace; color: rgb(215, 186, 125); background: none 0% 0% / auto repeat scroll padding-box border-box rgba(255, 255, 255, 0.1); padding: 2px 4px; border-radius: 3px; word-break: break-word; font-size: 0.9em;">training</code> phase</strong> (<a href="vscode-webview://13qe2srtq7rfa1e0j47ttranh8f06vadkqqo1rgicmqsrna8gimk/ecologits/impacts/training.py" target="_blank" rel="noopener noreferrer" style="color: rgb(55, 148, 255); text-decoration: rgb(55, 148, 255);">training.py</a>): final run, R&amp;D ×4, dataset storage, allocated per output token over the model's lifetime tokens. It requires a model release date (added to 146 models in <code style="font-family: monospace; color: rgb(215, 186, 125); background: none 0% 0% / auto repeat scroll padding-box border-box rgba(255, 255, 255, 0.1); padding: 2px 4px; border-radius: 3px; word-break: break-word; font-size: 0.9em;">models.json</code>) and a provider compute capacity (Anthropic, Google, Mistral, OpenAI). Otherwise a new <code style="font-family: monospace; color: rgb(215, 186, 125); background: none 0% 0% / auto repeat scroll padding-box border-box rgba(255, 255, 255, 0.1); padding: 2px 4px; border-radius: 3px; word-break: break-word; font-size: 0.9em;">training-not-modeled</code> warning is emitted. It is reported separately and <strong>not added to the totals</strong>, so existing numbers only shift by the network and building additions.</p><p style="white-space: pre-wrap; margin-top: 0.1em; margin-bottom: 0.2em; unicode-bidi: plaintext; color: rgb(204, 204, 204); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Roboto, sans-serif; font-size: 13px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(37, 37, 38); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Also fixed a real bug on the way: <code style="font-family: monospace; color: rgb(215, 186, 125); background: none 0% 0% / auto repeat scroll padding-box border-box rgba(255, 255, 255, 0.1); padding: 2px 4px; border-radius: 3px; word-break: break-word; font-size: 0.9em;">RangeValue.__rtruediv__</code> computed the wrong interval.</p><p style="white-space: pre-wrap; margin-top: 0.1em; margin-bottom: 0.2em; unicode-bidi: plaintext; color: rgb(204, 204, 204); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Roboto, sans-serif; font-size: 13px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(37, 37, 38); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Docs: new <a href="vscode-webview://13qe2srtq7rfa1e0j47ttranh8f06vadkqqo1rgicmqsrna8gimk/docs/methodology/llm_training.md" target="_blank" rel="noopener noreferrer" style="color: rgb(55, 148, 255); text-decoration: rgb(55, 148, 255);">llm_training.md</a>, updated <a href="vscode-webview://13qe2srtq7rfa1e0j47ttranh8f06vadkqqo1rgicmqsrna8gimk/docs/methodology/llm_inference.md" target="_blank" rel="noopener noreferrer" style="color: rgb(55, 148, 255); text-decoration: rgb(55, 148, 255);">llm_inference.md</a>, tutorial, warnings page, nav, and a <a href="vscode-webview://13qe2srtq7rfa1e0j47ttranh8f06vadkqqo1rgicmqsrna8gimk/docs/blog/posts/2026_09_impactia_integration.md" target="_blank" rel="noopener noreferrer" style="color: rgb(55, 148, 255); text-decoration: rgb(55, 148, 255);">blog post</a> whose byline is a placeholder (<code style="font-family: monospace; color: rgb(215, 186, 125); background: none 0% 0% / auto repeat scroll padding-box border-box rgba(255, 255, 255, 0.1); padding: 2px 4px; border-radius: 3px; word-break: break-word; font-size: 0.9em;">samuelrince</code>).</p><h2 aria-level="5" style="unicode-bidi: plaintext; color: rgb(204, 204, 204); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Roboto, sans-serif; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(37, 37, 38); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Something to report back to the ImpactIA team</h2><p style="white-space: pre-wrap; margin-top: 0.1em; margin-bottom: 0.2em; unicode-bidi: plaintext; color: rgb(204, 204, 204); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Roboto, sans-serif; font-size: 13px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(37, 37, 38); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Their spreadsheet has a unit inconsistency in cell <code style="font-family: monospace; color: rgb(215, 186, 125); background: none 0% 0% / auto repeat scroll padding-box border-box rgba(255, 255, 255, 0.1); padding: 2px 4px; border-radius: 3px; word-break: break-word; font-size: 0.9em;">Calcul!N115</code>: provider compute capacity is stored in kW but multiplied by FLOPS per W without the ×1000. This divides the estimated lifetime token volume by 1000 and is what produces their headline "training = 13% median". I implemented the guide's method with consistent units, so the share EcoLogits reports is much smaller:</p>
Model | Training share of GWP | Energy change, 100k input tokens
-- | -- | --
gpt-4.1-mini | 0.19% | +5.9%
claude-opus-4-6 | 0.04% | +15.2%
mistral-large-latest | 7.4% | +4.0%
gemini-2.5-flash | 0.01% | +10.3%

<p style="white-space: pre-wrap; margin-top: 0.1em; margin-bottom: 0.2em; unicode-bidi: plaintext; color: rgb(204, 204, 204); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Roboto, sans-serif; font-size: 13px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(37, 37, 38); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">The token-volume estimate remains the weak point of the method in both directions, which the methodology page states plainly. All parameters can be overridden, including <code style="font-family: monospace; color: rgb(215, 186, 125); background: none 0% 0% / auto repeat scroll padding-box border-box rgba(255, 255, 255, 0.1); padding: 2px 4px; border-radius: 3px; word-break: break-word; font-size: 0.9em;">model_lifetime_output_token_count</code>.</p><h2 aria-level="5" style="unicode-bidi: plaintext; color: rgb(204, 204, 204); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Roboto, sans-serif; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(37, 37, 38); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Deliberately left out</h2><ul style="padding-inline-start: 2em; color: rgb(204, 204, 204); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Roboto, sans-serif; font-size: 13px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(37, 37, 38); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><li style="unicode-bidi: plaintext;"><strong>RAG embedding stage</strong>: worth doing as an embeddings methodology, but needs embedding models in the repository and tracer support for embedding endpoints.</li><li style="unicode-bidi: plaintext;"><strong>"Front applicatif"</strong>: SNCF-specific hosting assumption, out of scope for an API-level library.</li><li style="unicode-bidi: plaintext;"><strong>AWARE-weighted embodied water</strong>: a different indicator from WCF in liters, so I did not merge it.</li><li style="unicode-bidi: plaintext;">ResilioDB embodied factors: kept EcoLogits' sourced values.</li></ul><h2 aria-level="5" style="unicode-bidi: plaintext; color: rgb(204, 204, 204); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Roboto, sans-serif; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(37, 37, 38); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Points needing your decision</h2><ul style="padding-inline-start: 2em; color: rgb(204, 204, 204); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Roboto, sans-serif; font-size: 13px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(37, 37, 38); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><li style="unicode-bidi: plaintext;"><strong>License</strong>: ImpactIA's guide and spreadsheet are CC BY-NC-SA. I re-implemented the formulas and wrote original documentation text with attribution, but you may want a legal check before merging into an MPL-2.0 project.</li><li style="unicode-bidi: plaintext;">Release dates are missing for a few models (<code style="font-family: monospace; color: rgb(215, 186, 125); background: none 0% 0% / auto repeat scroll padding-box border-box rgba(255, 255, 255, 0.1); padding: 2px 4px; border-radius: 3px; word-break: break-word; font-size: 0.9em;">claude-opus-4-7</code>, some Gemini 3.1 previews, rolling<span> </span><code style="font-family: monospace; color: rgb(215, 186, 125); background: none 0% 0% / auto repeat scroll padding-box border-box rgba(255, 255, 255, 0.1); padding: 2px 4px; border-radius: 3px; word-break: break-word; font-size: 0.9em;">-latest</code><span> </span>aliases), so they get the warning instead of a training estimate.</li><li style="unicode-bidi: plaintext;">Nothing is committed; review the diff and commit when satisfied.</li></ul>

### CODE REVIEW

**Comment by benoit-cty** [comment]
ccusage measure:
| Agent                │ Models          │      Input │     Output │  Cache Create │   Cache Read │  Total Tokens │  Cost (USD) │
| - Claude             │ - fable-5-1     │        996 │    144,004 │       430,255 │    8,806,508 │     9,381,763 │      $18.02 │

**Comment by samuelrince** [comment]
Hey @benoit-cty, thanks I think I'd rather meet them and discuss some parts of the methodology before reviewing this PR!

**Comment by XavierVerne** [comment]
Thanks for this _detailed_ pull request and all the work behind. A first few things on what you explained : 
1. Please let us know if there is a blocking condition regarding our licence, so as for us all to find a way. The *purpose* was to facilitate external integration while preventing people to close further improvments.
2. thanks for the bug, fixing in progress.
3. I understood that Samuel and Rosendo and Thibaud are about to talk to each other. May be you could join in ? refer back to Samuel I don't have all infos. This could accelerate things.
